Add can_view_editor_onboarding property to editor session events

dcalhoun opened the pull request:
* **Gutenberg:** https://github.com/WordPress/gutenberg/pull/32739
* **Gutenberg Mobile:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3625
* **WPAndroid:** https://github.com/wordpress-mobile/WordPress-Android/pull/14766

Add `can_view_editor_onboarding` property to editor session events to allow analyzing how capabilities like the editor onboarding impact user behavior. 

To test: 

1. Open Tracks to observe editor session events.
1. Uninstall any preexisting WordPress app from device. 
1. Install WordPress app for this PR. 
1. Launch the Gutenberg block editor. 
1. ℹ️ **Expected:** The "Tap to add content" tooltip is displayed. 
1. Close the Gutenberg block editor. 
1. Launch the Gutenberg block editor. 
1. ℹ️ **Expected:** The "Tap to add content" tooltip is _not_ displayed. 
1. Close the Gutenberg block editor. 
1. ℹ️ **Expected:** Four editor session events are captured: 
    * `wpios_editor_session_start` with `can_view_editor_onboarding: true`.
    * `wpios_editor_session_end` with `can_view_editor_onboarding: true`.
    * `wpios_editor_session_start` with `can_view_editor_onboarding: true`.
    * `wpios_editor_session_end` with `can_view_editor_onboarding: true`. 

## Regression Notes
1. Potential unintended areas of impact
  Modifications to the post editor analytics and publish ability could introduce regressions to publishing or editor-related analytic events. 
2. What I did to test those areas of impact (or what existing automated tests I relied on)
  Manually verified publishing posts worked and related analytic events were triggered for the Gutenberg block editor. 
3. What automated tests I added (or what prevented me from doing so)
  Additional automated tests for this single, experimental property addition are likely not that valuable. 

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dcalhoun committed:
Add capabilities property to editor_session_start event

Property added to allow analyzing how capabilities like the editor
onboarding impact user behavior.

guarani commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -823,11 +823,16 @@ extension GutenbergViewController: GutenbergBridgeDelegate {
     func gutenbergDidMount(unsupportedBlockNames: [String]) {
         if !editorSession.started {
+            var capabilities: [String: Bool] = [:]
+            for ability in self.gutenbergCapabilities() {
```
nitpick: I think you could use something like `for (key, value) in gutenbergCapabilities()` here.

dcalhoun replied:
Love it! Give me all the nitpicks you have, because I'm not as familiar with Swift. 😄

Commit:
Add editorCapabilities method to editors

dcalhoun committed:
Fix lint violation

dcalhoun committed:
Add capabilities property to editor_session_end event

Property added to allow analyzing how capabilities like the editor
onboarding impact user behavior.

dcalhoun committed:
Align editor session start and end invocations

Reduce code by utilizing the same approach for retrieving editor
capabilities.

dcalhoun committed:
Include single capability with boolean value

Sending all capabilities as a stringified JSON object is not all that
useful in our analytics system. Sending a single
`can_view_editor_onboarding` value allows segmenting by that value.

dcalhoun commented on WordPress/Classes/ViewRelated/Post/PostEditorAnalyticsSession.swift:
```diff
@@ -40,6 +40,11 @@ struct PostEditorAnalyticsSession {
             properties[Property.unsupportedBlocks] = blocksJSON
         }
+
+        if let canViewEditorOnboarding = capabilities["canViewEditorOnboarding"] {
```
Is there a better approach than using a string literal to access this key?

guarani replied:
I don't think there's a better way. Converting the capabilities dictionary back to `[Capabilities: Bool]` wouldn't help much.

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-iOS into update/editor-session-analytic-properties

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-iOS into update/editor-session-analytic-properties

dcalhoun committed:
Separate computation of onboarding visibility and analytic property

Computing these two values separately enables limiting display of the
editor onboarding to a single time while still allowing us to segment
related analytic events by users who have been included in the cohort.

dcalhoun commented:
@guarani this work has been refactored as we discussed in chat. Will you please review at your convenience?

guarani commented on WordPress/Classes/Utility/Editor/GutenbergSettings.swift:
```diff
@@ -178,14 +178,23 @@ class GutenbergSettings {
     func setHasLaunchedGutenbergEditor(_ hasLaunched: Bool) {
         database.set(hasLaunched, forKey: Key.hasLaunchedGutenbergEditor)
     }
 
```
It looks like `setHasLaunchedGutenbergEditor` isn't used anymore.

dcalhoun replied:
Good catch! I will remove this unused function.

dcalhoun replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/16574/commits/26b24a7078bbed8707ab77919362febb60dffd75.

guarani commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -827,10 +827,10 @@ extension GutenbergViewController: GutenbergBridgeDelegate {
             // If you need to refactor this, please ensure that the startup_time_ms property
             // is still reflecting the actual startup time of the editor
-            editorSession.start(unsupportedBlocks: unsupportedBlockNames)
+            editorSession.start(unsupportedBlocks: unsupportedBlockNames, canViewEditorOnboarding: self.canViewEditorOnboarding())
```
```suggestion
            editorSession.start(unsupportedBlocks: unsupportedBlockNames, canViewEditorOnboarding: canViewEditorOnboarding())
```

guarani replied:
While `self` can be used here, it's not needed. It's only normally needed inside closures.

dcalhoun replied:
Ah, OK. Thanks for the tip. I'll update all the relevant calls to remove `self`.

guarani commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -1052,10 +1052,14 @@ extension GutenbergViewController: GutenbergBridgeDataSource {
             // (https://github.com/wordpress-mobile/gutenberg-mobile/issues/3457) in self-hosted sites is fixed
             .reusableBlock: isWPComSite,
-            .canViewEditorOnboarding: gutenbergSettings.canViewEditorOnboarding()
+            .canViewEditorOnboarding: self.canViewEditorOnboarding() && !gutenbergSettings.hasLaunchedGutenbergEditor
```
```suggestion
            .canViewEditorOnboarding: canViewEditorOnboarding() && !gutenbergSettings.hasLaunchedGutenbergEditor
```

guarani commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -1052,10 +1052,14 @@ extension GutenbergViewController: GutenbergBridgeDataSource {
         ]
     }
 
+    func canViewEditorOnboarding() -> Bool {
+        gutenbergSettings.canViewEditorOnboarding()
+    }
+
```
nitpick: I'm fine leaving this in, but accessing `gutenbergSettings.canViewEditorOnboarding()` directly would seem enough, especially since this doesn't provide any more context (e.g. a more descriptive method name).

guarani replied:
Ignore that, I remember now that this method is needed because it's part of the `PublishingEditor` protocol.

guarani commented on WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift:
```diff
@@ -310,7 +316,7 @@ extension PublishingEditor where Self: UIViewController {
         /// Otherwise, we'll show an Action Sheet with options.
         if post.shouldAttemptAutoUpload && post.canSave() {
-            editorSession.end(outcome: .cancel)
+            editorSession.end(outcome: .cancel, canViewEditorOnboarding: self.canViewEditorOnboarding())
```
```suggestion
            editorSession.end(outcome: .cancel, canViewEditorOnboarding: canViewEditorOnboarding())
```

guarani commented on WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift:
```diff
@@ -320,7 +326,7 @@ extension PublishingEditor where Self: UIViewController {
             showPostHasChangesAlert()
         } else {
-            editorSession.end(outcome: .cancel)
+            editorSession.end(outcome: .cancel, canViewEditorOnboarding: self.canViewEditorOnboarding())
```
```suggestion
            editorSession.end(outcome: .cancel, canViewEditorOnboarding: canViewEditorOnboarding())
```

guarani reviewed (commented):
Just left some minor code comments but the code is looking great! I also tested the Tracks events and they also show up as expected.

Since you mentioned offline that you'll be making some modifications, I'll hold off giving the final approve.

dcalhoun committed:
Remove unnecessary self references

Co-authored-by: Paul Von Schrottky <paul.von.schrottky@automattic.com>

dcalhoun committed:
Rename Gutenberg capability property for clarity

The capability property passed to Gutenberg is dependent upon both the
percentage of eligible users and whether the editor has been launched
before. Renaming the capability was done to better communicate that it
is not tied to a single factor and match the naming pattern of other
capabilities.

dcalhoun committed:
Remove unused code

This was utilized in a previous iteration.

dcalhoun committed:
Add note regarding the flags attachment to an app installation

The flag resets if the app is uninstalled.

dcalhoun committed:
Update gutenberg ref

dcalhoun commented:
@Guarani I believe I have addressed each of your great comments. Will you please review this PR again at your earliest convenience? Thank you! 🙇🏻

dcalhoun committed:
Trigger new build

dcalhoun committed:
Temporarily remove code for debugging purposes

This should be reinstated before merging this work. This is removed to
see if it has any impact on the tooltip not displaying for Atomic sites.

dcalhoun committed:
Revert "Temporarily remove code for debugging purposes"

This reverts commit a77a33a178a366ead0e878043bc2626b4726c01c.

dcalhoun committed:
Fix race condition with hasLaunchedGutenbergEditor flag

Setting hasLaunchedGutenbergEditor within the gutenbergDidMount hook
sometimes caused an inaccurate value to be shared with Gutenberg if the
capabilities properties were re-computed multiple times during the first
Gutenberg session. By setting this flag during viewWillDisappear, we
ensure that the flag is only true after the first time a user utilizes
the Gutenberg editor, rather than within the first session.

dcalhoun committed:
Invoke super within ViewController instance method override

We should adhere to documentation for this instance method to avoid
issues that may occur by not invoking the super method.

guarani reviewed (approved):
Tested and working well! Code looks great too! 🚢 :shipit:

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-iOS into update/editor-session-analytic-properties

dcalhoun committed:
Update gutenberg ref
